sway-core/src/declaration_engine/declaration_engine.rs

```diff
@@ -158,26 +135,78 @@ impl DeclarationEngine {
         self.slab.get(*index).expect_struct(span)
     }
 
-    fn de_add_monomorphized_struct_copy(
+    pub(crate) fn de_get_monomorphized_decl(
         &self,
         original_id: DeclarationId,
-        new_copy: TypedStructDeclaration,
-    ) {
-        let span = new_copy.span();
-        let new_id =
-            DeclarationId::new(self.slab.insert(DeclarationWrapper::Struct(new_copy)), span);
-        self.de_add_monomorphized_copy(original_id, new_id)
+        type_arguments: &Vec<TypeArgument>,
+        span: &Span,
+    ) -> Result<DeclarationWrapper, CompileError> {
+        for monomorphized_decl in self
+            .de_get_monomorphized_copies(original_id)
+            .iter()
+            .cloned()
+        {
+            if monomorphized_decl.type_parameters() == type_arguments {
```

Comment:
This line will cause the cache to fail in some cases (e.g. nested generic function calls).

For example this:

```rust
fn foo<T, F>(a: T, b: F) -> F {
    b
}

fn bar<T, F>(a: T, b: F) -> F {
    foo::<T, F>(a, b)
}
```

would generate a false cache hit due to the implementation of `PartialEq` on `TypeInfo::UnknownGeneric`: https://github.com/FuelLabs/sway/blob/a989e6ca0caf6ace4630cd72a2ff8c875c141a43/sway-core/src/type_system/type_info.rs#L205